README.md
# What is TeamSpeak 3?

TeamSpeak is a proprietary voice-over-Internet Protocol (VoIP) application for audio communication between users on a chat channel, much like a telephone conference call. Users typically use headphones with a microphone. The client software connects to a TeamSpeak server of the user's choice, from which the user may join chat channels.

# Quick Start

## Basic

```
docker run \
    -p 9987:9987/udp \
    -p 30033:30033 \
    -p 10011:10011 \
    hackebein/ts3server
```
# Environment

## Basic

### TS3SERVER_DB_CLEAR
Default: 0

Since: 3.0.0

> If set to "1", the server database will be cleared before starting up the server. This is mainly used for testing. Usually this parameter should not be specified, so all server settings will be restored when the server process is restarted.

### TS3SERVER_DB_CLIENT_DAYS
Default: 90

Since: 3.0.0

> Defines how many days to keep unused client identities. Auto-pruning is triggered on every start and on every new month while the server is running.

### TS3SERVER_LICENSE
Default: view

Since: 3.0.0

> if set to "accept", the server will assume you have read and accepted the license it comes with. If this is set to "view", the ts3server will not start.

### TS3SERVER_LOG_QUERY_COMMANDS
Default: 1

Since: 3.0.0

> If set to '1', the server will log every ServerQuery command executed by clients. This can be useful while trying to diagnose several different issues.

### TS3SERVER_PATCH_ENABLE
Default: false

Since: 3.0.0

> If set to 'true', the server patch routine will be enabled.

### TS3SERVER_PATCH_BADGES_DISABLE
Default: false

Since: 3.0.9

> If set to 'true', the badges of other users are hidden.

### TS3SERVER_PATCH_GDPR_SAVE
Default: false

Since: 3.0.0

> If set to 'true', the server save less ip adresses to fulfill the General Data Protection Regulation (GDPR).

### TS3SERVER_PRINT_ENV
Default: false

Since: 3.0.0

> If set to 'true', shows the environment at the beginning.

### TS3SERVER_QUERY_PASSWORD
Default:

Since: 3.0.0

> Defines the server query admin password.

## database

### TS3SERVER_DB_CONNECTIONS
Default: 10

Since: 3.0.0

> The number of database connections used by the server. Please note that changing this value can have an affect on your servers performance. Possible values are 2-100.

### TS3SERVER_DB_HOST
Default: 127.0.0.1

Since: 3.0.0

> The hostname or IP addresss of your MariaDB/MySQL server.

### TS3SERVER_DB_LOG_DAYS
Default: 90

Since: 3.0.0

> Defines how many days to keep database log entries. Auto-pruning is triggered on every start and on every new month while the server is running.

### TS3SERVER_DB_LOGGING_DISBALE
Default: 1

Since: 3.0.0

-

### TS3SERVER_DB_NAME
Default: test

Since: 3.0.0

> The name of a database on your MariaDB/MySQL server. Note that this database must be created before the TeamSpeak Server is started. Please use 'utf8mb4' character encoding for the database.

### TS3SERVER_DB_PASSWORD
Default:

Since: 3.0.0

> The password used to authenticate with your MariaDB/MySQL server.

### TS3SERVER_DB_PLUGIN
Default: ts3db_sqlite3

Since: 3.0.0

> Possible values: ts3db_sqlite3, ts3db_mariadb

### TS3SERVER_DB_PLUGINPARAMETER
Default: ts3db.ini

Since: 3.0.0

### TS3SERVER_DB_PORT
Default: 3306

Since: 3.0.0

> The TCP port of your MariaDB/MySQL server.

### TS3SERVER_DB_SOCKET
Default:

Since: 3.0.0

> The name of the Unix socket file to use, for connections made via a named pipe to a local server.

### TS3SERVER_DB_SQL_CREATE_PATH
Default: create_sqlite

Since: 3.0.0

> Possible values: create_sqlite, create_mariadb

### TS3SERVER_DB_SQL_PATH
Default: sql

Since: 3.0.0

> The physical path where your SQL script files are located.

### TS3SERVER_DB_UPDATE_DISABLE
Default: 0

Since: 3.0.0

> If set to '1', new permissions will not be added to existing groups automatically. Note that this can break your server configuration if you do not update them manually.

### TS3SERVER_DB_USER
Default: root

Since: 3.0.0

> The username used to authenticate with your MariaDB/MySQL server.

### TS3SERVER_DB_WAITUNTILREADY
Default: 30

Since: 3.1.0

## Special

### TS3SERVER_CRASHDUMPS
Default: crashdumps

Since: 3.6.0

> When the server crashes, a crashdump is created that may be send to teamspeak to help fixing the crash. The location where the crashdumps are saved too, can be changed with this parameter. This feature is currently not supported on FreeBSD and Alpine versions of the TeamSpeak Server.

### TS3SERVER_FILETRANSFER_IP
Default: 0.0.0.0

Since: 3.0.0

> Comma separated IP list which the file transfers are bound to.

### TS3SERVER_FILETRANSFER_PORT
Default: 30033

Since: 3.0.0

> TCP Port opened for file transfers.

### TS3SERVER_LICENSE_PATH
Default:

Since: 3.0.0

The physical path where your license file is located.

### TS3SERVER_LOG_APPEND
Default: 0

Since: 3.0.1

> If set to '1', the server will not create a new logfile on every start. Instead, the log output will be appended to the previous logfile. The logfile name will only contain the ID of the virtual server.

### TS3SERVER_LOG_PATH
Default: logs

Since: 3.0.0

> The physical path where the server will create logfiles.

### TS3SERVER_MACHINE_ID
Default:

Since: 3.0.0

> Optional name of this server process to identify a group of servers with the same ID. This can be useful when running multiple TeamSpeak Server instances on the same database. Please note that we strongly recommend that you do NOT run multiple server instances on the same SQLite database.

### TS3SERVER_PROXY
Default:

Since: 3.3.0

> If set, the server will use the specified proxy to contact the accounting server over http.
>
> Supported formats are:
> * domainname:port
> * ipv4address:port
> * [ipv6address]:port

### TS3SERVER_QUERY_BLACKLIST
Default: query_ip_blacklist.txt

Since: 3.0.0

### TS3SERVER_QUERY_BRUTFORCECHECK_DISABLE
Default: 0

Since: 3.0.8

> If set to "1", the server will skip and bruteforce protection for whitelisted Ip addresses for the ServerQuery interface.

### TS3SERVER_QUERY_BUFFER
Default: 20

Since: 3.3.0

> Server Query connections have a combined maximum buffer size. When this limit is exceeded, the connection using the most memory is closed. The default is 20, which means the maximum amount of buffered data is 20 megabyte. The minimum is 1 megabyte. Make sure to only enter positive integer numbers here.

### TS3SERVER_QUERY_DOCS_PATH
Default: serverquerydocs/

Since: 3.3.0

> Physical location where the server is looking for the documents used for the help command in ServerQuery.

### TS3SERVER_QUERY_RAW_ENABLE
Default: true

Since: 3.3.0

### TS3SERVER_QUERY_RAW_IP
Default: 0.0.0.0

Since: 3.0.0

### TS3SERVER_QUERY_RAW_PORT
Default: 10011

Since: 3.0.0

### TS3SERVER_QUERY_SSH_ENABLE
Default: false

Since: 3.3.0

### TS3SERVER_QUERY_SSH_IP
Default: 0.0.0.0

Since: 3.3.0

### TS3SERVER_QUERY_SSH_PORT
Default: 10022

Since: 3.3.0

### TS3SERVER_QUERY_SSH_RSA_HOST_KEY
Default: ssh_host_rsa_key

Since: 3.3.0

> The physical path to the ssh_host_rsa_key to be used by query. If it does not exist, it will be created when the server is starting up.

### TS3SERVER_QUERY_WHITELIST
Default: query_ip_whitelist.txt

Since: 3.0.0

### TS3SERVER_VOICE_DEFAULT_CREATE
Default: 1

Since: 3.0.0

> Normally one virtual server is created automatically when the TeamSpeak Server process is started. To disable this behavior, set this parameter to '0'. In this case you have to start virtual servers manually using the ServerQuery interface.

### TS3SERVER_VOICE_DEFAULT_PORT
Default: 9987

Since: 3.0.0

> UDP port open for clients to connect to. This port is used by the first virtual server, subsequently started virtual servers will open on increasing port numbers.

### TS3SERVER_VOICE_IP
Default: 0.0.0.0

Since: 3.0.0

> Comma separated IP list on which the server instance will listen for incoming voice connections.


releases
3.0.0|amd64|b1b33603badeddde51718c813845bf1dd338857b304e517d7dcbce4b13c415bf|http://dl.4players.de/ts/releases/3.0.0/teamspeak3-server_linux-amd64-3.0.0.tar.gz
3.0.1|amd64|c04ed2b878e4c84fbd62b5f8b22276cd036503218b4bc3487f852810e468f71e|http://dl.4players.de/ts/releases/3.0.1/teamspeak3-server_linux-amd64-3.0.1.tar.gz
3.0.2|amd64|4ec8e74e8f479fdcb187e5c920ee5fa165daa5678ddcc1caf6483c138a5e24f9|http://dl.4players.de/ts/releases/3.0.2/teamspeak3-server_linux-amd64-3.0.2.tar.gz
3.0.3|amd64|6b825b6d80866adbb115a5ae334ca12d2c5b99c2d99bcfae4479c16f4bfb3c60|http://dl.4players.de/ts/releases/3.0.3/teamspeak3-server_linux-amd64-3.0.3.tar.gz
3.0.5|amd64|61f158d445262dc682676b17c86c36db2008b9e472401bd6409a6f8dde3d595a|http://dl.4players.de/ts/releases/3.0.5/teamspeak3-server_linux-amd64-3.0.5.tar.gz
3.0.6|amd64|c755c4c79f8dc16f00b13a60efe7ab78eab667a0dcd299c25f3e4f728a5cd946|http://dl.4players.de/ts/releases/3.0.6/teamspeak3-server_linux-amd64-3.0.6.tar.gz
3.0.6.1|amd64|5d222d1f6b28db856788e0004a1a2143372c82fecbd0651e8a575cee8742ef3c|http://dl.4players.de/ts/releases/3.0.6.1/teamspeak3-server_linux-amd64-3.0.6.1.tar.gz
3.0.7|amd64|ac81b21e135407cb862716ee4dc56f316e1e7e5e7a91141dc3fd2775142b3c44|http://dl.4players.de/ts/releases/3.0.7/teamspeak3-server_linux-amd64-3.0.7.tar.gz
3.0.7.1|amd64|500b9a75dbe5287db7f4d8b788e53b73069c9d1d49cebc8b6b5e0b3081e0473b|http://dl.4players.de/ts/releases/3.0.7.1/teamspeak3-server_linux-amd64-3.0.7.1.tar.gz
3.0.7.2|amd64|1be5cb2dd03168c8c2b8900748439545e122d0cdbf66e2c9bf3d9bed1d9553f6|http://dl.4players.de/ts/releases/3.0.7.2/teamspeak3-server_linux-amd64-3.0.7.2.tar.gz
3.0.8|amd64|f40bd7976a6fb084868f57bf10e5d954265a732914f45b396aedda530fd9c0c8|http://dl.4players.de/ts/releases/3.0.8/teamspeak3-server_linux-amd64-3.0.8.tar.gz
3.0.9|amd64|3559624ce21142a90c21c8a8e789d420b6043cc97698713a070cddccd8918b08|http://dl.4players.de/ts/releases/3.0.9/teamspeak3-server_linux-amd64-3.0.9.tar.gz
3.0.10|amd64|4144bbdc92ea765bda4a7dfbb176482f56caa8bf52531d9d8afc41401518335b|http://dl.4players.de/ts/releases/3.0.10/teamspeak3-server_linux-amd64-3.0.10.tar.gz
3.0.10.1|amd64|db650269c6521aabdc2673a3ad76780bd455c9f1fc3e1081277473a3452e662c|http://dl.4players.de/ts/releases/3.0.10.1/teamspeak3-server_linux-amd64-3.0.10.1.tar.gz
3.0.10.2|amd64|c65bcc5f28da74a5a5a1545b92e2295ab5e6bdfd40940a91241d2c094777fa69|http://dl.4players.de/ts/releases/3.0.10.2/teamspeak3-server_linux-amd64-3.0.10.2.tar.gz
3.0.10.3|amd64|9606dd5c0c3677881b1aab833cb99f4f12ba08cc77ef4a97e9e282d9e10b0702|http://dl.4players.de/ts/releases/3.0.10.3/teamspeak3-server_linux-amd64-3.0.10.3.tar.gz
3.0.11|amd64|990441aada1e9306e6152773278544901db2e6bb4c74ab0c2ca9254ee77b4519|http://dl.4players.de/ts/releases/3.0.11/teamspeak3-server_linux-amd64-3.0.11.tar.gz
3.0.11.1|amd64|84807a15f45b8a578d28781f56fe5da72a8794f82d0ed0629e50565fad671f08|http://dl.4players.de/ts/releases/3.0.11.1/teamspeak3-server_linux-amd64-3.0.11.1.tar.gz
3.0.11.2|amd64|0aca07461b2711782347c1186258eed54cae8b9b14a2364acfe8f15053cf10ab|http://dl.4players.de/ts/releases/3.0.11.2/teamspeak3-server_linux-amd64-3.0.11.2.tar.gz
3.0.11.3|amd64|448d35de0de5c0c9b17feba0697fc232920fb68ac2880bbbf022e0386a17b63b|http://dl.4players.de/ts/releases/3.0.11.3/teamspeak3-server_linux-amd64-3.0.11.3.tar.gz
3.0.11.4|amd64|92123aff892740c88acc30a6e3e4df1615be44f9780acd396d663bac91323a40|http://dl.4players.de/ts/releases/3.0.11.4/teamspeak3-server_linux-amd64-3.0.11.4.tar.gz
3.0.12|amd64|b6f1afc0c8f677af2e1694e99a8ca2475b99b85c9bf6652fd0f96ad2256e2769|http://dl.4players.de/ts/releases/3.0.12/teamspeak3-server_linux_amd64-3.0.12.tar.bz2
3.0.12.1|amd64|99d1382e748fe2099d519592c65bc4923b58b70fec2091b09a47e0a4ab34ff33|http://dl.4players.de/ts/releases/3.0.12.1/teamspeak3-server_linux_amd64-3.0.12.1.tar.bz2
3.0.12.2|amd64|24180a1a4c2e7d33b0574fa6b9860f25c248b96638410c01622c275e7284184a|http://dl.4players.de/ts/releases/3.0.12.2/teamspeak3-server_linux_amd64-3.0.12.2.tar.bz2
3.0.12.3|amd64|07f72ae51bde9bb1f144455b5e352f5cd8e22da112ec265509a50e2b9af86b41|http://dl.4players.de/ts/releases/3.0.12.3/teamspeak3-server_linux_amd64-3.0.12.3.tar.bz2
3.0.12.4|amd64|6bb0e8c8974fa5739b90e1806687128342b3ab36510944f576942e67df7a1bd9|http://dl.4players.de/ts/releases/3.0.12.4/teamspeak3-server_linux_amd64-3.0.12.4.tar.bz2
3.0.13|amd64|93d702c30e5f3745ad156b020a751e448afa64459d8769d6d848a3ef43b28de9|http://dl.4players.de/ts/releases/3.0.13/teamspeak3-server_linux_amd64-3.0.13.tar.bz2
3.0.13.1|amd64|81aebe641a1292b274fd77ee493a6ce2c77ccd8d1fff1baf5fc502ec77e9760c|http://dl.4players.de/ts/releases/3.0.13.1/teamspeak3-server_linux_amd64-3.0.13.1.tar.bz2
3.0.13.2|amd64|f7dd49a38126fb6a4c0e7818575c7d898556330cafe5fc59b20d9f10744f6799|http://dl.4players.de/ts/releases/3.0.13.2/teamspeak3-server_linux_amd64-3.0.13.2.tar.bz2
3.0.13.3|amd64|e9f48c8a9bad75165e3a7c9d9f6b18639fd8aba63adaaa40aebd8114166273ae|http://dl.4players.de/ts/releases/3.0.13.3/teamspeak3-server_linux_amd64-3.0.13.3.tar.bz2
3.0.13.4|amd64|cff353c3f395175ba251c787c5ca1cbb3d339be1ae1afe0cf10216a6e81ae5af|http://dl.4players.de/ts/releases/3.0.13.4/teamspeak3-server_linux_amd64-3.0.13.4.tar.bz2
3.0.13.5|amd64|bd5933dd17d17f93d56f69332927cd1ce6f34439ec464a0ce2ca73102d85080c|http://dl.4players.de/ts/releases/3.0.13.5/teamspeak3-server_linux_amd64-3.0.13.5.tar.bz2
3.0.13.6|amd64|19ccd8db5427758d972a864b70d4a1263ebb9628fcc42c3de75ba87de105d179|http://dl.4players.de/ts/releases/3.0.13.6/teamspeak3-server_linux_amd64-3.0.13.6.tar.bz2
3.0.13.7|amd64|b5b142c70859efdaffe50ec0a94ce54b32b3f232507fd7a52457c961a00f8b0d|http://dl.4players.de/ts/releases/3.0.13.7/teamspeak3-server_linux_amd64-3.0.13.7.tar.bz2
3.0.13.8|amd64|460c771bf58c9a49b4be2c677652f21896b98a021d7fff286e59679b3f987a59|http://dl.4players.de/ts/releases/3.0.13.8/teamspeak3-server_linux_amd64-3.0.13.8.tar.bz2

ts3server_3.0.3/ts3server_startscript.sh
#!/usr/bin/env bash
set -eu
cd "$(dirname "$(readlink -f "$0")")"

# vercomp '3.0.0' '3.1.0'
# case $? in
	# 0) foo='=';;
	# 1) foo='>';;
	# 2) foo='<';;
# esac
vercomp () {
    if [[ $1 == $2 ]]; then
        return 0
    fi
    local IFS=.
    local i ver1=($1) ver2=($2)
    # fill empty fields in ver1 with zeros
    for ((i=${#ver1[@]}; i<${#ver2[@]}; i++)); do
        ver1[i]=0
    done
    for ((i=0; i<${#ver1[@]}; i++)); do
        if [[ -z ${ver2[i]} ]]; then
            # fill empty fields in ver2 with zeros
            ver2[i]=0
        fi
        if ((10#${ver1[i]} > 10#${ver2[i]})); then
            return 1
        fi
        if ((10#${ver1[i]} < 10#${ver2[i]})); then
            return 2
        fi
    done
    return 0
}

vercompare () {
	vercomp $1 $3
	case $? in
		0) op='=';;
		1) op='>';;
		2) op='<';;
	esac

	if [[ $2 = *"${op}"*  ]]; then
		echo true
	else
		echo false
	fi
}

# environment
export LD_LIBRARY_PATH="/usr/local/lib"
export TS3SERVER_VERSION=${TS3SERVER_VERSION:-0}
export TS3SERVER_LICENSEPATH=${TS3SERVER_LICENSEPATH:-}
export TS3SERVER_IP_WHITELIST=${TS3SERVER_IP_WHITELIST:-query_ip_whitelist.txt}
export TS3SERVER_IP_BLACKLIST=${TS3SERVER_IP_BLACKLIST:-query_ip_blacklist.txt}
export TS3SERVER_LOG_PATH=${TS3SERVER_LOG_PATH:-/app/logs}
export TS3SERVER_LOG_QUERY_COMMANDS=${TS3SERVER_LOG_QUERY_COMMANDS:-0}
export TS3SERVER_LOG_APPEND=${TS3SERVER_LOG_APPEND:-0}
export TS3SERVER_DB_PLUGIN=${TS3SERVER_DB_PLUGIN:-ts3db_sqlite3}
export TS3SERVER_DB_PLUGINPARAMETER=${TS3SERVER_DB_PLUGINPARAMETER:-/app/ts3db.ini}
export TS3SERVER_DB_SQLPATH=${TS3SERVER_DB_SQLPATH:-/app/sql/}
export TS3SERVER_DB_SQLCREATEPATH=${TS3SERVER_DB_SQLCREATEPATH:-create_sqlite}
export TS3SERVER_DB_CONNECTIONS=${TS3SERVER_DB_CONNECTIONS:-10}
export TS3SERVER_DB_CLIENTKEEPDAYS=${TS3SERVER_DB_CLIENTKEEPDAYS:-30}
export TS3SERVER_DB_HOST=${TS3SERVER_DB_HOST:-}
export TS3SERVER_DB_PORT=${TS3SERVER_DB_PORT:-3306}
export TS3SERVER_DB_USER=${TS3SERVER_DB_USER:-}
export TS3SERVER_DB_PASSWORD=${TS3SERVER_DB_PASSWORD:-}
export TS3SERVER_DB_NAME=${TS3SERVER_DB_NAME:-}
export TS3SERVER_DB_SOCKET=${TS3SERVER_DB_SOCKET:-}
export TS3SERVER_DB_WAITUNTILREADY=${TS3SERVER_DB_WAITUNTILREADY:-30}
export TS3SERVER_QUERY_PASSWORD=${TS3SERVER_QUERY_PASSWORD:-}
export TS3SERVER_PATCH_ENABLE=${TS3SERVER_PATCH_ENABLE:-false}
export TS3SERVER_PATCH_BADGES_DISABLE=${TS3SERVER_PATCH_BADGES_DISABLE:-false}
export TS3SERVER_SUPPORT_BADGE=$(vercompare ${TS3SERVER_VERSION} '>=' '3.0.9')

if [[ ${TS3SERVER_SUPPORT_BADGE} != true ]]; then
	TS3SERVER_PATCH_BADGES_DISABLE=false
fi

echo ---
printenv | grep '^TS3SERVER_' | grep -v 'PASSWORD' | sort
echo ---

# main
if [ -e ts3server.dist ]; then
	cp -a ts3server.dist ts3server
	rm ts3server.dist
fi
if [[ "${TS3SERVER_PATCH_ENABLE}" == "true" ]]; then
	cp -a ts3server ts3server.dist
	# Patch for disable badges
	if [[ "${TS3SERVER_PATCH_BADGES_DISABLE}" == "true" ]]; then
		sed -e 's/client_badges/client_BADGES/g' -i ts3server
	fi
fi

# TODO: cmd param
# query_port=10011
# filetransfer_ip=0.0.0.0 # do also if port only is set!
# filetransfer_port=30033

cat <<- EOF >/app/ts3server.ini
licensepath=${TS3SERVER_LICENSEPATH}
query_ip_whitelist=${TS3SERVER_IP_WHITELIST}
query_ip_blacklist=${TS3SERVER_IP_BLACKLIST}
dbplugin=${TS3SERVER_DB_PLUGIN}
dbpluginparameter=${TS3SERVER_DB_PLUGINPARAMETER}
dbsqlpath=${TS3SERVER_DB_SQLPATH}
dbsqlcreatepath=${TS3SERVER_DB_SQLCREATEPATH}
dbconnections=${TS3SERVER_DB_CONNECTIONS}
dbclientkeepdays=${TS3SERVER_DB_CLIENTKEEPDAYS}
logpath=${TS3SERVER_LOG_PATH}
logquerycommands=${TS3SERVER_LOG_QUERY_COMMANDS}
logappend=${TS3SERVER_LOG_APPEND}
EOF
cat <<- EOF >/app/ts3db.ini
[config]
host='${TS3SERVER_DB_HOST}'
port='${TS3SERVER_DB_PORT}'
username='${TS3SERVER_DB_USER}'
password='${TS3SERVER_DB_PASSWORD}'
database='${TS3SERVER_DB_NAME}'
socket=${TS3SERVER_DB_SOCKET}
wait_until_ready='${TS3SERVER_DB_WAITUNTILREADY}'
EOF

# prepare execution
set -- "$@" inifile=/app/ts3server.ini
if [ -n "${TS3SERVER_QUERY_PASSWORD}" ]; then
    set -- "$@" serveradmin_password=${TS3SERVER_QUERY_PASSWORD}
fi

# execution
exec ./ts3server "$@"
